scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryClient.scala

```diff
@@ -351,52 +342,30 @@ class BigQueryClient private (private val projectId: String,
     }
   }
 
-  private def temporaryTable: TableReference = {
+  private def temporaryTable(location: String): TableReference = {
     val now = Instant.now().toString(TIME_FORMATTER)
     val tableId = TABLE_PREFIX + "_" + now + "_" + Random.nextInt(Int.MaxValue)
     new TableReference()
       .setProjectId(projectId)
-      .setDatasetId(BigQueryClient.stagingDataset)
+      .setDatasetId(STAGING_DATASET_PREFIX + location.toLowerCase)
       .setTableId(tableId)
   }
 
-  private def createJobReference(projectId: String, jobIdPrefix: String): JobReference = {
-    val fullJobId = projectId + "-" + UUID.randomUUID().toString
-    new JobReference().setProjectId(projectId).setJobId(fullJobId)
-  }
-
-  private def createJobConfigurationQuery(sqlQuery: String,
-                                          destinationTable: TableReference,
-                                          flattenResults: Boolean,
-                                          useLegacySql: Boolean): JobConfigurationQuery =
-    new JobConfigurationQuery()
-      .setQuery(sqlQuery)
-      .setUseLegacySql(useLegacySql)
-      .setAllowLargeResults(true)
-      .setFlattenResults(flattenResults)
-      .setPriority(PRIORITY)
-      .setCreateDisposition("CREATE_IF_NEEDED")
-      .setWriteDisposition("WRITE_EMPTY")
-      .setDestinationTable(destinationTable)
-
   private def delayedQueryJob(sqlQuery: String,
                               destinationTable: TableReference,
                               flattenResults: Boolean): QueryJob = new QueryJob {
     override def waitForResult(): Unit = self.waitForJobs(this)
     override lazy val jobReference: Option[JobReference] = {
-      prepareStagingDataset()
+      val location = extractLocation(sqlQuery)
+      prepareStagingDataset(location)
       val isLegacy = isLegacySql(sqlQuery, flattenResults)
       if (isLegacy) {
         logger.info(s"Executing legacy query: `$sqlQuery`")
       } else {
         logger.info(s"Executing SQL query: `$sqlQuery`")
       }
-      val queryConfig = createJobConfigurationQuery(
-        sqlQuery, destinationTable, flattenResults, isLegacy)
-      val jobConfig = new JobConfiguration().setQuery(queryConfig)
-      val jobReference = createJobReference(projectId, JOB_ID_PREFIX)
-      val job = new Job().setConfiguration(jobConfig).setJobReference(jobReference)
-      Some(bigquery.jobs().insert(projectId, job).execute().getJobReference)
+      val tryRun = runQuery(sqlQuery, destinationTable, flattenResults, isLegacy, dryRun = false)
+      Some(tryRun.get.getJobReference)
```

Comment:
Yes at this point the query is guaranteed to be correct since it passed `isLegacySql`. Any error should be unexpected and propagated.